File: 3306-count-of-substrings-containing-every-vowel-and-k-consonants-ii/3306-count-of-substrings-containing-every-vowel-and-k-consonants-ii.java
class Solution {
    public long countOfSubstrings(String word, int k) {
        var n = word.length();
        Map<Character, Integer> vowels = new HashMap<>();
        for (var c : "aeiou".toCharArray()) vowels.put(c, 0);
        int v = 0, c = 0;   // count of distinct vowels and count of consonants
        long ans = 0; 

        // find pos of first consonant on right of each index
        var nextConsonant = new int[n];
        var lastConsonant = n;
        for(var i=n-1; i>=0; i--){
            nextConsonant[i] = lastConsonant;
            if(!isVowel(word.charAt(i))) lastConsonant = i;
        }

        for(int l=0, r=0; r < n; r++){
            // expand window on right
            var ch = word.charAt(r);
            if(isVowel(ch)){
                vowels.put(ch, vowels.get(ch) + 1);
                if(vowels.get(ch) == 1) v++;    // found new vowel
            }
            else c++;

            // shrink window on left
            while(c > k){
                var ch1 = word.charAt(l++);
                if(isVowel(ch1)){
                    vowels.put(ch1, vowels.get(ch1) - 1);
                    if(vowels.get(ch1) == 0) v--;    // lost a vowel
                }
                else c--;
            }

            // count matching strings
            while(v == 5 && c == k){
                ans += nextConsonant[r] - r;
                var ch1 = word.charAt(l++);
                if(isVowel(ch1)){
                    vowels.put(ch1, vowels.get(ch1) - 1);
                    if(vowels.get(ch1) == 0) v--;    // lost a vowel
                }
                else c--;
            }
        }

        return ans;
    }

    private boolean isVowel(char ch){
        return (ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u');
    }
}

File: 3306-count-of-substrings-containing-every-vowel-and-k-consonants-ii/3306-count-of-substrings-containing-every-vowel-and-k-consonants-ii.py
class Solution:
    def countOfSubstrings(self, word: str, k: int) -> int:
        n = len(word)
        vowels = {'a': 0, 'e': 0, 'i': 0, 'o': 0, 'u': 0}
        v = c = ans = 0   # count of distinct vowels and count of consonants and ans

        # find pos of first consonant on right of each index
        nextConsonant = [0] * n
        lastConsonant = n
        for i in range(n-1, -1, -1):
            nextConsonant[i] = lastConsonant
            if not self.is_vowel(word[i]): lastConsonant = i

        l = 0
        for r in range(0, n):
            # expand window on right
            ch = word[r]
            if self.is_vowel(ch):
                vowels[ch] += 1
                if vowels[ch] == 1: v += 1    # found new vowel
            else: c += 1

            # shrink window on left
            while c > k:
                ch1 = word[l]
                if self.is_vowel(ch1):
                    vowels[ch1] -= 1
                    if vowels[ch1] == 0: v -= 1    # lost a vowel
                else: c -= 1
                l += 1

            # count matching strings
            while v == 5 and c == k:
                ans += nextConsonant[r] - r
                ch1 = word[l]
                if self.is_vowel(ch1):
                    vowels[ch1] -= 1
                    if vowels[ch1] == 0: v -= 1    # lost a vowel
                else: c -= 1
                l += 1

        return ans

    def is_vowel(self, ch: str) -> bool:
        return ch in {'a', 'e', 'i', 'o', 'u'}